Implemented new smoothing brush algorithms

sts1skj opened the pull request:
For the position-smoothing brush, the new algorithm is a "balanced pair circle fit" smoother.  It finds balanced pairs of adjacent points and fits a circle through each pair.

For the mask-smoothing and weight-smoothing brushes, the new algorithms are "balanced pair parabola fit" smoothers.  They find balanced pairs of adjacent points and fit a parabola through each pair.

I think the circle fitter works well.  The parabola fitter doesn't work great, it seems to me, but is still probably better than the existing hclap code.

sts1skj committed:
Fixed some brush unwelding and normal bugs
- Fixed bug in AABBTree::AABBTreeNode::IntersectSphere that would cause it
to return a bad distance to the triangle.  I added Triangle::DistanceToPoint
to correctly calculate this distance.  This bug would make
TweakBrush::queryPoints often pick the wrong triangle as the center of its
update with the Connected-Only brush option on.  This, in turn, would cause
Mesh::ConnectedPointsInSphere to misbehave, causing brushes to unweld
vertices.  It might also sometimes cause brushes to have no effect on
some vertices that they should effect.
- Fixed bug in Mesh::ConnectedPointsInSphere that would cause it to often miss
welded vertices.  This would cause the Connected-Only brush option to
sometimes make brushes unweld vertices.
- Simplified Mesh::ConnectedPointsInSphere a little.
- Added Mesh::ConnectedPointsInTwoSpheres for use by brushes with the
"NeedMirrorMergedQuery" flag.  This fixes a bug that would rarely cause
points to not get adjusted.
- Fixed flaw in Mesh::SmoothNormals that would cause triangles to sometimes
flicker black.  The problem was that it would zero the normals before
calculating them.  I made it do the calculation in a temporary copy of the
normals.
- Fixed bug in Mesh::SmoothNormals that would cause it to produce bad
results for welded vertices in a partial update.  ("!= 0" should have been
"== 0").  This would make triangles flicker near welded vertices when using
a brush with live normal updates.
- Fixed flaw in live-normals updating where it wouldn't update the normals
of vertices adjacent to a changed vertex, despite the fact that those
vertices have changed normals too.  This flaw would cause triangles to
look wrong along the edge of the area affected by the brush until the brush
stroke ended.
- Used new return-value version of trinormal several places.

sts1skj committed:
Implemented new smoothing brush algorithms
For the position-smoothing brush, the new algorithm is a "balanced pair
circle fit" smoother.  It finds balanced pairs of adjacent points and
fits a circle through each pair.

For the mask-smoothing and weight-smoothing brushes, the new algorithms
are "balanced pair parabola fit" smoothers.  They find balanced pairs
of adjacent points and fit a parabola through each pair.

sts1skj commented:
Some questions:
1. Should the user be able to pick which smoothing algorithm they want?  Currently they can't.  (I suggest no.)
2. Should using balanced pairs be a tool option that applies to any of the three smoothers?  Any of the existing lap or hclap smoothers could use balanced pairs as an option.  (I suggest no.)
3. The new smoothers use a lap-style update instead of an hclap-style update.  (hclap-style updating tugs the results back towards the starting values, sort of.)  Should I rewrite them to use hclap-style updating?  (I suggest no.)
4. I put the functions ParabolaFit, CircleFitMidpoint, and CircleFitNearestPoint in TweakBrush.cpp.  These functions could reasonably be moved elsewhere.  ParabolaFit relies on Mesh functions, so it could be easily transferred into Mesh.{h,cpp}, but moving it to nifly's Object3d.{h,cpp} would be more messy, though definitely doable.  CircleFitMidpoint and CircleFitNearestPoint could easily be moved to Mesh.{h,cpp} or Object3d.{h,cpp}.  Are these functions okay in TweakBrush.cpp, or should they be moved?
5. The new mask and weight smoothers are not great.  Should they be dropped in favor of the existing hclap smoothers?

sts1skj commented:
One thing I noticed while experimenting with the new bpcf position-smoother is that it sometimes produces slightly poor results near points with connections like this:
![image](https://user-images.githubusercontent.com/59774938/155627288-c0cec403-5b45-4783-93a9-c2b70c1aa2ca.png)
The reason is that the bpcf code relies on normals calculated by Mesh::SmoothNormals, and that function doesn't produce great results when a point's triangles are unbalanced on one side, like the point in the image above.  It produces a normal for the central point that is more dominated by the upper triangles' normals than the lower triangles' normals, because there are four upper triangles and only two lower triangles.  So the normals at points like this aren't quite right.  I'm not sure, but I think I've noticed this a few times while playing Skyrim.

So maybe Mesh::SmoothNormals could be improved.  One idea is to do a plane fit through balanced pairs of neighboring vertices.  Another idea is calculating triangle normals for a set of temporarily generated triangles that only use balanced pairs of neighboring vertices (instead of using the mesh's own neighboring triangles).  Should I look into this?

sts1skj committed:
Show mirror of brush circle with x-mirror on

sts1skj committed:
Merge branch 'dev' into smoothAlg

ousnius committed:
Make mirror circle transparent instead of a darker red

ousnius commented:
@sts1skj There's been some changes in dev (adjacent points) that require a rebase/merge.

sts1skj committed:
Merge remote-tracking branch 'origin/dev' into smoothAlg

sts1skj committed:
Merge remote-tracking branch 'origin/dev' into smoothAlg

sts1skj committed:
Update nifly (undoing an accidental reversion)

sts1skj committed:
Use new precomputed vertex adjacency lists
This commit is mostly a simplification of existing code to take advantage
of the new Mesh::adjVerts.  But it also fixes a problem in BuildAdjacency
that would make it include a vertex's welds in its adjacency list.

sts1skj committed:
Merge remote-tracking branch 'mygithub/showMirrorBrush' into smoothAlg

sts1skj committed:
Merge remote-tracking branch 'mygithub/smoothAlg' into smoothAlg

sts1skj committed:
Convenience functions for weldVerts
LeastWeldedVertexIndex, DoForEachWeldedVertex, GetWeldSet

sts1skj committed:
More tweaking of vertex adjacencies
- Rewrote BuildVertexAdjacency again so that hopefully now welds
are included properly.  I tried to write this function so that it
is crystal clear what it does, so hopefully we won't make any more
mistakes reading the code and maintaining it.
- Deleted function Mesh::GetOneVertexNormal.  I added this a few weeks ago,
and I now realize it does nothing useful.
- Used the new weldVerts convenience functions in a couple of more places.
- Added more comments to Mesh.h to hopefully improve clarity and
maintainability of adjVerts, BuildVertexAdjacency, and GetAdjacentPoints.
- Changed MaskMore and MaskLess so that they first check to see if any
point's welds have a different mask value.  If so, they fix that first.
- Made MaskMore and MaskLess undoable.
- Moved MaskMore and MaskLess from OutfitStudio.h to OutfitStudio.cpp, as
these functions are getting big enough and complicated enough that I felt
it was no longer appropriate for them to be in a header.
- Made mask changes in wxGLPanel::SelectVertex undoable.

sts1skj committed:
Made InvertMask, OnMaskWeighted, and OnMaskBoneWeighted undoable

ousnius committed:
Revert merge error b78721b

ousnius committed:
Make "Clear Mask" undoable
